### PR TITLE
Podcast ownership join table (MLH Fellowship #81)

### PR DESCRIPTION
**Pull request by Manasa2850**
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [x] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Related Tickets & Documents

https://github.com/MLH-Fellowship/dev.to/issues/81

## QA Instructions, Screenshots, Recordings

_Please replace this line with instructions on how to test your changes, as well
as any relevant images for UI changes._

## Added tests?

- [x] Yes
- [ ] No, and this is why: the code I've added doesn't require tests as of now
- [ ] I need help with writing tests

## Added to documentation?

- [ ] [Developer Docs](https://docs.forem.com) and/or [Admin Guide](https://forem.gitbook.io/forem-admin-guide/)
- [ ] README
- [x] No documentation needed

## [optional] Are there any post deployment tasks we need to perform?
No

## [optional] What gif best describes this PR or how it makes you feel?

alt_text

### CODE REVIEW

**Comment by Manasa2850** [comment]
@benhalpern I've made all the changes suggested by Chris. 
Could you please review it and let me know if I need to make some changes?

File: app/controllers/podcast_ownerships_controller.rb
Comment (rhymes) [inline app/controllers/podcast_ownerships_controller.rb:9]: where is `podcast_ownership_params` defined?

File: app/controllers/podcast_ownerships_controller.rb
Comment (rhymes) [inline app/controllers/podcast_ownerships_controller.rb:32]: It's probably better to add `@podcast_ownership.errors_as_sentence` to the error message, this way you can tell the client which errors are present:

https://github.com/forem/forem/blob/910f79e04e402214f33600cddb44367e9c6e5d45/app/models/application_record.rb#L56-L58

File: app/controllers/podcast_ownerships_controller.rb
Comment (rhymes) [inline app/controllers/podcast_ownerships_controller.rb:1]: This controller doesn't have neither tests nor templates, I wonder if it should be added in its own separate PR, after we merge the table (it seems also off topic, as mentioned by the title)

File: app/models/podcast.rb
Comment (rhymes) [inline app/models/podcast.rb:5]: is the goal to have multiple owners on a podcast? It might be beneficial to call this `.owners` instead of `.users`, similar to how we call `.creator` instead of `.user` the creator of the podcast

File: app/models/podcast_ownership.rb
Comment (rhymes) [inline app/models/podcast_ownership.rb:4]: Same here, you could do something like:

```ruby
belongs_to :owner, class_name: "User", inverse_of: :owners
belongs_to :podcast, inverse_of: :podcasts
```

File: app/policies/podcast_ownership_policy.rb
Comment (rhymes) [inline app/policies/podcast_ownership_policy.rb:7]: probably should be the same as `create?`, forbidding this action to banned users

File: app/policies/podcast_ownership_policy.rb
Comment (rhymes) [inline app/policies/podcast_ownership_policy.rb:3]: should we check `user_is_banned?` here as well?

File: app/policies/podcast_ownership_policy.rb
Comment (rhymes) [inline app/policies/podcast_ownership_policy.rb:21]: what's the scenario for it to be something other than a `PodcastOwnership` ?

File: db/migrate/20201114130315_create_podcast_ownerships.rb
Comment (rhymes) [inline db/migrate/20201114130315_create_podcast_ownerships.rb:5]: can you add `null: false` here? Unfortunately by default [add_reference](https://api.rubyonrails.org/classes/ActiveRecord/ConnectionAdapters/SchemaStatements.html#method-i-add_reference) defaults to `null: true`

**Review by rhymes — CHANGES_REQUESTED** [review]
Hi @Manasa2850, welcome to the Forem repo! Thank you for the contribution.

I'm unfamiliar with this feature at the moment, would you mind filling in the rest of the pull request template? This way I can help review it better. Right now I just checked the code as is and left some notes

**Comment by rhymes** [comment]
@Manasa2850 given the conflicts I suspect you haven't updated your local branch before submitting this, can you also please take a look at that, I suggest solving the conflict before addressing any other change

**Comment by Manasa2850** [comment]
Hi @rhymes. Thanks for the review!
This is the issue I'm working on https://github.com/MLH-Fellowship/dev.to/issues/81

File: app/controllers/podcast_ownerships_controller.rb
Comment (Manasa2850) [reply app/controllers/podcast_ownerships_controller.rb:9]: PodcastOwnership is the join table between users and podcasts so we get the params from there

File: app/controllers/podcast_ownerships_controller.rb
Comment (Manasa2850) [reply app/controllers/podcast_ownerships_controller.rb:1]: Hmm yeah.. Chris suggested that I add a Pundit policy and a controller too...hence I added it here...we can have it in a separate PR if required too!

**Comment by rhymes** [comment]
@Manasa2850 thanks for the details!

**Comment by Manasa2850** [comment]
@rhymes I'll resolve the merge conflicts and make the changes you've suggested!

File: app/policies/podcast_ownership_policy.rb
Comment (Manasa2850) [reply app/policies/podcast_ownership_policy.rb:21]: The user who tries to edit the podcast need not be the owner. So if it's not an instance of PodcastOwnership (i.e the user is not the owner of the podcast), they will not be permitted to modify it.

File: app/controllers/podcast_ownerships_controller.rb
Comment (Manasa2850) [reply app/controllers/podcast_ownerships_controller.rb:1]: @rhymes I'll also be working on a follow up issue https://github.com/MLH-Fellowship/dev.to/issues/85
So I can remove the controller and the associated pundit policy from this PR if you wish

Comment (rhymes) [reply app/controllers/podcast_ownerships_controller.rb:1]: Yeah, I think it'd be simpler to merge one bit at the time, this way we don't merge broken code (if anyone accesses the controller URL they will get a 400 error because the templates aren't there)

File: app/controllers/podcast_ownerships_controller.rb
Comment (rhymes) [reply app/controllers/podcast_ownerships_controller.rb:9]: Yeah but the method `podcast_ownership_params` doesn't exist :) I guess you'll have to define it in a similar fashion to https://github.com/forem/forem/blob/910f79e04e402214f33600cddb44367e9c6e5d45/app/controllers/admin/users_controller.rb#L167-L175

but as suggested, I think this should all split in a separate PR :)

File: app/policies/podcast_ownership_policy.rb
Comment (rhymes) [reply app/policies/podcast_ownership_policy.rb:21]: I'm not sure I understand fully that it corresponds to the logic you wrote but once you'll write a test for the policy you can test its feasibility. See https://github.com/forem/forem/tree/master/spec/policies for policy specs

**Comment by rhymes** [comment]
@Manasa2850 I think you have quite a few issues derived by conflict resolution. 
For example https://github.com/forem/forem/pull/11410/files#diff-9802ca3c9c4cf89904fd44bc114e35ebdf2c5dd3d5b645491e2b253e1afef29bR119 contains something that shouldn't be there: `has_many :page_views, dependent: :destroy` is already present a few line below as `has_many :page_views, dependent: :nullify` which is the correct statement

The `schema.rb` also is completely off :-( check the diff here https://github.com/forem/forem/pull/11410/files#diff-cbf8b25d6d853acf58fa9057841f407d29ffe90649c75cf9e3f51b2d6e3aa1d3 

that's why Travis isn't working

File: app/controllers/podcast_ownerships_controller.rb
Comment (Manasa2850) [reply app/controllers/podcast_ownerships_controller.rb:9]: Okay @rhymes I'll remove the controller and the pundit policy files from this PR. We can keep this PR just for the join table.
Thanks for suggesting that!

**Comment by Manasa2850** [comment]
@rhymes I would like to create a new PR by rebasing and removing the controller and pundit policy. I'll be closing this one.
Hope that's fine :)

**Comment by Manasa2850** [comment]
@rhymes I've resolved the merge conflicts!
I've also removed the controller and pundit policy files from this PR. I'll add it in another PR
Please let me know if something more needs to be done

**Comment by Manasa2850** [comment]
I'm not sure why the `<<<<<<< HEAD` line is coming. Sorry I'm a little new to resolving merge conflicts through the command line

File: app/models/podcast.rb
Comment (rhymes) [inline app/models/podcast.rb:4]: this is already present on line 7

File: app/models/podcast.rb
Comment (rhymes) [reply app/models/podcast.rb:5]: see my note above about explicitly naming this `.owners`

File: app/models/podcast_ownership.rb
Comment (rhymes) [reply app/models/podcast_ownership.rb:4]: see my note above about renaming this to `owner` if possible

File: app/models/user.rb
Comment (rhymes) [inline app/models/user.rb:123]: as https://github.com/MLH-Fellowship/dev.to/issues/82 is coming down the pipeline, the association here should be named `:owned_podcasts`, so that a user can be the owner of one or more podcasts and also appear in one or more podcasts

File: db/migrate/20201114130315_create_podcast_ownerships.rb
Comment (rhymes) [reply db/migrate/20201114130315_create_podcast_ownerships.rb:5]: don't forget `null: false`

**Review by rhymes — CHANGES_REQUESTED** [review]
Hi @Manasa2850, there a few details to adjust but we're almost there :) See my notes below, thank you!

File: app/models/user.rb
Comment (Manasa2850) [reply app/models/user.rb:123]: @rhymes can we check this once with @benhalpern before changing it? Because in my previous meeting with Ben and Chris, we had decided on `podcast_ownerships`

**Comment by Manasa2850** [comment]
@rhymes I'm getting this error in the Travis CI test `ActiveRecord::PendingMigrationError:` 
Could you please help me resolve this?

**Comment by Manasa2850** [comment]
@rhymes I ran the command `rails db:migrate RAILS_ENV=test` like mentioned in the Travis CI log. But this changes my `db/schema.rb` completely which then leads to merge conflicts :(

File: app/models/user.rb
Comment (rhymes) [reply app/models/user.rb:123]: @Manasa2850 what I meant is changing `user.podcasts` to `user.owned_podcasts` or `user.podcasts_owned`

that's because https://github.com/forem/forem/pull/11354 will add the concept of "podcasts that I have been a guest on" which is different from "podcasts that I owned and manage". They can't be both called `podcasts`. The name `through :podcast_ownerships` is correct :)

File: app/models/podcast_ownership.rb
Comment (rhymes) [inline app/models/podcast_ownership.rb:5]: I believe this has been mutuated from https://github.com/forem/forem/pull/11354/files#diff-80ed95ceccb910e831a5c371a28a8b35b37729d2d4ad1d063702c9a084086b1fR4-R5 but this model doesn't have the `role` column, and it shouldn't have.

The uniqueness validation if needed needs to be backed by a DB index, see https://github.com/forem/forem/pull/11354/files#diff-1953bb445767352adbf509787151f41c6f293305f5b4d65189ccb33a15a261c3R11

ps. the name of the column is `user_id`, `:owner` it's just an alias to help us humans in the application logic :), if you check [belongs_to](https://api.rubyonrails.org/classes/ActiveRecord/Associations/ClassMethods.html#method-i-belongs_to) options section you can find how `belongs_to :owner, class_name: "User"` works

**Review by rhymes — CHANGES_REQUESTED** [review]
@Manasa2850 AFAIK the `schema.rb` in this PR doesn't correspond to what's written in `db/migrate/20201114130315_create_podcast_ownerships.rb`:

```ruby
    t.bigint "podcast_id"
    t.bigint "user_id"
```

are missing `null: false`

and the version of the schema.rb file is wrong. In the diff is `2020_11_10_180614` but the migration file you added is `db/migrate/20201114130315_create_podcast_ownerships.rb` so the version should be `2020_11_14_130315`, which is why Travis is saying that a migration is pending, because the schema is one migration behind and thus it thinks there's still one to be run

ps. don't forget the model tests (also run them locally if possible so you can anticipate Travis errors) as explained in my previous review here https://github.com/forem/forem/pull/11410#pullrequestreview-530771966

**Comment by Manasa2850** [comment]
@rhymes I get this error in Travis CI `Factory not registered: "podcast_ownership"` in the spec file. I'm a little new to writing tests. Could you please help me out with this?

**Comment by rhymes** [comment]
Sure @Manasa2850, as `podcast_ownership` is a new class it needs a factory for tests. A factory is a set of instructions that builds a real model object, through a shortcut. If you look into https://github.com/forem/forem/tree/master/spec/factories you'll find one per each model. What Travis is telling you is that `FactoryBot` (the gem we use to build objects in tests) is not aware of `podcast_ownership`, thus you just need a new `spec/factories/podcast_ownership.rb` file (no need to explicitly register anything) which contains something similar to:

https://github.com/forem/forem/blob/994fbf99ad5f60f6a1ff7db2f7f874aa4b186340/spec/factories/rating_votes.rb#L1-L3

For example, I'd imagine, since podcast ownership needs a user and a podcast, you could do:

```ruby
FactoryBot.define do
  factory :podcast_ownership do
    owner { user }
    podcast
  end
end
```

(I haven't tested it, check [FactoryBot's documentation](https://github.com/thoughtbot/factory_bot/blob/master/GETTING_STARTED.md#specifying-the-factory))

`FactoryBot` when inside a test you then call `create(:podcast_ownership)` will load or create a user object and a podcast object automatically.

Anyway, that won't work until the definition of the various associations is fixed as some are incomplete. If you load the console and try to create a `PodcastOwnership` object you will see:

```ruby
[4] pry(main)> po = PodcastOwnership.create!(owner: User.last, podcast: Podcast.last)
  User Load (0.9ms)  SELECT "users".* FROM "users" ORDER BY "users"."id" DESC LIMIT $1  [["LIMIT", 1]]
  Podcast Load (0.3ms)  SELECT "podcasts".* FROM "podcasts" ORDER BY "podcasts"."id" DESC LIMIT $1  [["LIMIT", 1]]
ActiveModel::MissingAttributeError: can't write unknown attribute `owner_id`
from /Users/rhymes/.gem/ruby/2.7.2/gems/activemodel-6.0.3.4/lib/active_model/attribute.rb:206:in `with_value_from_database'
```

see the `:foreign_key` section in Rails's [belongs_to](https://api.rubyonrails.org/classes/ActiveRecord/Associations/ClassMethods.html#method-i-belongs_to) documentation.

I suggest you load the console and play with creating `PodcastOwnership` objects manually, see if then you can retrieve them from `user.owned_podcasts` and `podcast.owners`

Good job so far!

**Comment by Manasa2850** [comment]
@rhymes I added the factory. I also tested it out on rails console!
There's a merge conflict regarding the `schema.rb` version. If I rebase that, I'll get the `ActiveRecord::PendingMigrationError:`
Could you let me know what I need to do?

**Comment by Manasa2850** [comment]
@rhymes could you help me with the Travis CI error? `Expected PodcastOwnership to have a belongs_to association called owner (PodcastOwnership does not have a owner_id foreign key.)`
When I had written `it { is_expected.to belong_to(:owner) }` in the spec file, I got the error `Expected PodcastOwnership to have a belongs_to association called user (no association called user)`

**Comment by rhymes** [comment]
Hi @Manasa2850, I'll take a look, would it be okay with you if I went ahead and adjusted the relations with a commit?

**Comment by Manasa2850** [comment]
@rhymes. Sure you can do that. Thank you

**Comment by rhymes** [comment]
@Manasa2850 I can't push to your branch, but I exported the patch file in a gist: https://gist.githubusercontent.com/rhymes/98c74c5c190508105ca717f0406719b9/raw/78820d192c0394325c56ea89d6c8fcd866bdbf36/0001-Adjust-various-many-to-many-relations.patch

If you download it you can apply it locally with:

```shell
$ git am 0001-Adjust-various-many-to-many-relations.patch
$ rm 0001-Adjust-various-many-to-many-relations.patch
```

and then you can push it here with `git push`

File: db/schema.rb
Comment (rhymes) [inline db/schema.rb:1275]: this needs to be manually reverted to the value it has on the left

**Review by rhymes — COMMENTED** [review]
Aside from the patch and the `schema.rb` fix there's one last thing that we need to take care of before we merge this in: uniqueness.

Right now a user can be the owner of a podcast multiple times, but I'm sure we want one user to appear as the owner of a podcast only once. How can we fix that? By adding the combination of a unique index on both `podcast_id` and `user_id` in the `podcast_ownerships` table and a validation for such uniqueness in the model file.

You can take inspiration from https://github.com/forem/forem/blob/cf619caf70f5ed129303b5b701b6e438913ed350/app/models/article.rb#L66 and https://github.com/forem/forem/pull/11354/files#diff-1953bb445767352adbf509787151f41c6f293305f5b4d65189ccb33a15a261c3R11 for example



Sorry that wasn't made clear before hand.

**Comment by Manasa2850** [comment]
@rhymes I've added everything you've suggested. I guess I need to make some changes to `spec/services/users/delete_spec.rb` too because that's causing the Travis CI build to fail. Could you let me know what changes I need to make there? Thanks!

**Comment by rhymes** [comment]
@Manasa2850 yeah, I can look into that. Thanks for adding the index in the migration, there's one last step: https://github.com/forem/forem/pull/11410/files#diff-82c6591538e42fcabb8338269551c7fee28225c07759c053023d25844609194dR8 is missing the corresponding index in the schema (you need to rollback the migration and re-run it so that it picks up the new index)

**Comment by rhymes** [comment]
@Manasa2850 the test should be fixed by changing this in the `app/models/user.rb` on line 121:

```ruby
  has_many :podcast_ownerships, dependent: :destroy, inverse_of: :owner
```

what's missing is the indication of the inverse of that relation

**Comment by rhymes** [comment]
Hi @Manasa2850, to solve the following:

```text
  1) PodcastOwnership validations is expected to validate that :podcast_id is case-sensitively unique within the scope of :user_id

     Failure/Error: it { is_expected.to validate_uniqueness_of(:podcast_id).scoped_to(:user_id) }

     

     ActiveRecord::InvalidForeignKey:

       PG::ForeignKeyViolation: ERROR:  insert or update on table "podcast_ownerships" violates foreign key constraint "fk_rails_3710d65292"

       DETAIL:  Key (podcast_id)=(0) is not present in table "podcasts".
```

you can add this to the test:

```ruby
  describe "validations" do
    subject { podcast_ownership }
```

By default rspec tries to create an empty object, but that's not going to work because of the foreign key constraint as mentioned in the error, but if we tell it to recycle an already valid object, then the error goes away.

**Comment by Manasa2850** [comment]
@rhymes thank you! I'll do that!

File: db/schema.rb
Comment (rhymes) [inline db/schema.rb:887]: I don't this is correct @Manasa2850, it doesn't reflect what's written in `db/migrate/20201114130315_create_podcast_ownerships.rb`

If you run the revert the migration and re-run it you'll notice the following changes from this version:

```diff
diff --git a/db/schema.rb b/db/schema.rb
index 526190a79..00ddfdfa7 100644
--- a/db/schema.rb
+++ b/db/schema.rb
@@ -880,11 +880,12 @@
 
   create_table "podcast_ownerships", force: :cascade do |t|
     t.datetime "created_at", precision: 6, null: false
-    t.bigint "podcast_id"
+    t.bigint "podcast_id", null: false
     t.datetime "updated_at", precision: 6, null: false
-    t.bigint "user_id"
-    t.index ["podcast_id"], name: "index_podcast_ownerships_on_podcast_id", unique: true
-    t.index ["user_id"], name: "index_podcast_ownerships_on_user_id", unique: true
+    t.bigint "user_id", null: false
+    t.index ["podcast_id", "user_id"], name: "index_podcast_ownerships_on_podcast_id_and_user_id", unique: true
+    t.index ["podcast_id"], name: "index_podcast_ownerships_on_podcast_id"
+    t.index ["user_id"], name: "index_podcast_ownerships_on_user_id"
   end
```

A tip: now that we have a unique index on both columns, we don't need the individual indexes on each column, see `index: true` in the migation file. 

Could you address that?

Basically you need to run:

- `bundle exec rails db:rollback` or `rails db:migrate:down VERSION=20201114130315`
- remove the `index: true` from both columns in the migration file
- re-run `bundle exec rails db:migrate`
- check the changes to the schema and commit those that pertain to `podcast_ownerships`
- push the code

Let me know if anything isn't clear

**Review by rhymes — CHANGES_REQUESTED** [review]

File: db/schema.rb
Comment (Manasa2850) [reply db/schema.rb]: @rhymes I ran the migration again!

Comment (rhymes) [reply db/schema.rb]: > A tip: now that we have a unique index on both columns, we don't need the individual indexes on each column, see index: true in the migation file.

Can you also remove the `index: true` from the singular fields? See my instructions above

Comment (rhymes) [reply db/schema.rb]: @Manasa2850 don't forget to re-run the migration after having removed `index: true`

Comment (Manasa2850) [reply db/schema.rb]: @rhymes yup I've done that!

Comment (rhymes) [reply db/schema.rb]: These two lines tell me that the migration still contain separate indexes for those columns:

```ruby
    t.index ["podcast_id"], name: "index_podcast_ownerships_on_podcast_id"
    t.index ["user_id"], name: "index_podcast_ownerships_on_user_id"
```

If you check the documentation for [references/add_reference](https://api.rubyonrails.org/classes/ActiveRecord/ConnectionAdapters/SchemaStatements.html#method-i-add_reference) you'll notice how `true` is the default for the `index` param. I guess it needs to be set as `false`. Could you address that?

Comment (Manasa2850) [reply db/schema.rb]: @rhymes running `rails db:migrate` was still not reflecting the change :(
So I had to remove those 2 lines manually

Comment (rhymes) [reply db/schema.rb]: @Manasa2850 that's because you need to set explicitly `index: false` in the migration, see my comment above. The migration file and the schema need to reflect each other, they can't be misaligned :)

**Review by rhymes — APPROVED** [review]
Thanks for the changes @Manasa2850! 

LGTM

**Comment by Manasa2850** [comment]
@rhymes thanks a ton for all the help!

**Review by mstruve — APPROVED** [review]
Welcome to Forem @Manasa2850! I love how clean and concise this code is. I always cringe when PRs try to do too much at once but you really kept this nice and contained which made reviewing it a piece of cake. 

Hope to see you tackling other issues in the future!

**Review by Ridhwana — APPROVED** [review]

**Pull request merged**
